Vectorize tensor lerp kernel

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84845

Fixes #86964

In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

peterbell10 committed:
Vectorize tensor lerp kernel

In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

peterbell10 committed:
Update on "Vectorize tensor lerp kernel"


In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

peterbell10 committed:
Update on "Vectorize tensor lerp kernel"


In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

peterbell10 committed:
Update on "Vectorize tensor lerp kernel"


In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

peterbell10 committed:
Update on "Vectorize tensor lerp kernel"


In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

peterbell10 committed:
Update on "Vectorize tensor lerp kernel"


In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Vectorize tensor lerp kernel"


In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

lezcano commented on aten/src/ATen/native/cpu/LerpKernel.cpp:
```diff
@@ -4,18 +4,68 @@
+  auto mask = is_lerp_weight_small(weight);
+  auto coeff = vec_t::blendv(weight - vec_t(1), weight, mask);
+  auto base = vec_t::blendv(end, start, mask);
+  return vec::fmadd(coeff, end - start, base);
```
That's more like it!

lezcano reviewed (approved):
Great stuff!

peterbell10 committed:
Update on "Vectorize tensor lerp kernel"


Fixes #86964

In a simple timeit benchmark I see 1.7x speedup for complex64, from 6.7 us to
3.9 us; and a 3.2x speedup for float32, from 6.2 us to 1.9 us.

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
